### DISCUSSION TITLE
Tracing static binaries

### DISCUSSION THREAD

**Discussion by ghost** [bpftrace/bpftrace#2184] [General]
Hi all,

As a newcomer to eBPF ecosystem I stumbled on one practical issue that I am not able to find an explanation for. I have  a specific need to trace a static binary, namely I have idea to trace some functions with `uprobe` and `uretprobe`. Unfortunately the specific mention of the binary I have is statically linked which does not seem to work. There are no probes available if I list with `bpftrace -l 'uprobe:path/to/binary:*'`.

I have another version of binary which is the usual dynamically linked binary. But I cannot really use it for this specific task. I can just test that my approach to facing user-space functions works.

If it makes any sense, the binary program I am talking about  is static build of ffmpeg as provided from https://johnvansickle.com/ffmpeg/.

Do I understand it correctly that static binaries cannot be traced with uprobes?

**Comment by brendangregg** [comment]
No. Check your symbols. uprobes can trace arbitrary instruction offsets, so it's not a problem with uprobes.

**Reply by ghost** [reply]
So the absence of symbols is the culprit, right?

```
$ objdump -tT /usr/local/bin/ffmpeg

/usr/local/bin/ffmpeg:     file format elf64-x86-64

objdump: /usr/local/bin/ffmpeg: not a dynamic object
SYMBOL TABLE:
no symbols


DYNAMIC SYMBOL TABLE:
no symbols
```

**Reply by ghost** [reply]
Seems in this specific case there isn't anything I can do if I have no way to get the symbols. Dead end. :cry:

**Reply by fbs** [reply]
You can still attach to the address of the function, if you know it. But building ffmpeg yourself is probably the better option

**Reply by ghost** [reply]
@fbs yeah, I do not know if I can figure out the address of function. I mean I know the name of function and that usually in non-striped dynamic `ffmpeg` case it is part of `libavcodec` dynamically loaded library.

Surely building ffmpeg myself is one approach albeit it is going to be rather painful and time consuming to try and reproduce the build. If I go the route of building ffmpeg, then I am in the realm of patching source code and adding "traces" (either logging or some other form of `printf()`) I need - thus rendering `bpftrace` approach redundant in my case.

Well, ok then.
